misc/dreambooth_app.py

```diff
@@ -0,0 +1,331 @@
+# and from that account you'll need to accept the model license [here](https://huggingface.co/CompVis/stable-diffusion-v1-4).
+#
+# Lastly, you'll need to create a token from that account and share it with Modal
+# under the name `"huggingface"`. Follow the instructions [here](https://modal.com/secrets).
+#
+# Then, you can kick off a training job with
+# `modal app run dreambooth_app.py --function-name train`
+# It should take about five minutes.
+
+
+@stub.function(
+    image=image,
+    gpu=gpu,  # finetuning is VRAM hungry, so this should be an A100
+    cpu=8,  # request enough CPUs to feed the GPU
+    shared_volumes={
+        str(MODEL_DIR): volume,
+    },
+    timeout=600,  # 10 minutes
+    secrets=[modal.Secret.from_name("huggingface")],
+    interactive=True,
+)
+def train(config=TrainConfig()):
+    import subprocess
+
+    from accelerate.utils import write_basic_config
+    import huggingface_hub
+    from smart_open import open
+
+    # set up runner-local image and shared model weight directories
+    img_path = load_images(config.images_file_url)
+    os.makedirs(MODEL_DIR, exist_ok=True)
+
+    # set up hugging face accelerate library for fast training
+    write_basic_config(mixed_precision="fp16")
+
+    # authenticate to hugging face so we can download the model weights
+    hf_key = os.environ["HUGGINGFACE_TOKEN"]
+    huggingface_hub.login(hf_key)
+
+    # fetch the training script from Hugging Face's GitHub repo
+    raw_repo_url = "https://raw.githubusercontent.com/huggingface/diffusers"
+    script_commit_hash = "30220905c4319e46e114cf7dc8047d94eca226f7"
+    script_path = "examples/dreambooth/train_dreambooth.py"
+    script_url = f"{raw_repo_url}/{script_commit_hash}/{script_path}"
+
+    with open(script_url) as from_file:
+        script_content = from_file.readlines()
+        with open("train_dreambooth.py", "w") as to_file:
+            to_file.writelines(script_content)
+
+    # define the training prompt
+    instance_phrase = f"{config.instance_name} {config.class_name}"
+    prompt = f"{config.prefix} {instance_phrase} {config.postfix}".strip()
+
+    # run training -- see huggingface accelerate docs for details
+    subprocess.run(
+        [
+            "accelerate",
+            "launch",
+            "train_dreambooth.py",
+            "--pretrained_model_name_or_path=CompVis/stable-diffusion-v1-4",
+            f"--instance_data_dir={img_path}",
+            f"--output_dir={MODEL_DIR}",
+            f"--instance_prompt='{prompt}'",
+            f"--resolution={config.resolution}",
+            f"--train_batch_size={config.train_batch_size}",
+            f"--gradient_accumulation_steps={config.gradient_accumulation_steps}",
+            f"--learning_rate={config.learning_rate}",
+            f"--lr_scheduler={config.lr_scheduler}",
+            f"--lr_warmup_steps={config.lr_warmup_steps}",
+            f"--max_train_steps={config.max_train_steps}",
+        ]
+    )
```

Comment:
might be good to add something like `stdout=sys.stdout, stderr=sys.stderr, check=True` to `subprocess.run` in case the process has any issues

Reply:
Wait, by default the subprocess inherits the stdout/stderr from the parent, and FWICT that works fine with `modal app run` here.

Also, noting here the  most of the work happens in `train_dreambooth.py`, which `accelerate` is also calling in a subprocess. It is called with `check=True`, so most errors are already handled. But adding `check` will catch issues at the level of the `accelerate` execution.

Reply:
> Wait, by default the subprocess inherits the stdout/stderr from the parent

Oh nice, I didn't realize that

```
>>> subprocess.run(["echo", "xyz"])
xyz
```